lib/quicly.c

```diff
@@ -3841,10 +3842,19 @@ static int do_send(quicly_conn_t *conn, quicly_send_context_t *s)
         assert(min_packets_to_send <= s->max_datagrams);
 
         if (restrict_sending) {
-            /* PTO (try to send new data when handshake is done, otherwise retire oldest handshake packets and retransmit) */
+            /* PTO: when handshake is in progress, send from the very first unacknowledged byte so as to maximize the chance of
+             * making progress. When handshake is complete, transmit new data if any, or retransmit the oldest unacknowledged
+             * data that is considered inflight. */
             QUICLY_PROBE(PTO, conn, conn->stash.now, conn->egress.sentmap.bytes_in_flight, conn->egress.cc.cwnd,
                          conn->egress.loss.pto_count);
-            if ((ret = mark_packets_on_pto(conn)) != 0)
+            size_t bytes_to_mark = min_packets_to_send * conn->egress.max_udp_payload_size;
+            if (conn->initial != NULL && (ret = mark_packets_on_pto(conn, QUICLY_EPOCH_INITIAL, &bytes_to_mark)) != 0)
+                goto Exit;
+            if (bytes_to_mark != 0 && conn->handshake != NULL &&
+                (ret = mark_packets_on_pto(conn, QUICLY_EPOCH_HANDSHAKE, &bytes_to_mark)) != 0)
+                goto Exit;
+            if (bytes_to_mark != 0 && !scheduler_can_send(conn) &&
```

Comment:
What does this condition `!scheduler_can_send(conn)` represent?
I got confused because "mark packets for retransmission, only when the scheduler does not let me send" does not make sense to me. Perhaps I'm missing some important background here.

Reply:
I think that is a good question.

`scheduler_can_send` is a wrapper function of the `quicly_stream_scheduler_t::can_send`. The function returns a boolean indicating if the scheduler has some data that can be sent immediately. The callback is used in pair with the `do_send` function. Reading your comment, I tend to agree that the name is confusing. Though, I am not sure if there's a better pair for the current names (i.e., "can_send" and "do_send").

Reply:
The idea is to send new data if possible. `scheduler_can_send()` returns `true` if the application has new data to send, in which case, we don't need to mark any packets as lost. If the application does not have any new data to send, then we mark packets as lost below using `mark_packets_on_pto()`.

This does assume that if there's _any_ new data to send, then we'll only send new data. I think that is fine, but it may be worth writing a comment to that effect here.

Reply:
I think a comment can help here if you don't want to change the name. Something like "If scheduler can send data, new application data is available. Do not mark already-sent packets if so."

Reply:
We already have a comment explaining what this code block does, and I think that covers what we do when there is new data or not. In acf1dcd, I have slightly changed the description hoping that it would be more precise.

Regarding the term "can_send", I think we should discuss the issue separately from this PR if necessary.

Reply:
@hfujita, proposed a comment.